lib/io/can.cpp

```diff
@@ -1,10 +1,121 @@
 #include "can.hpp"
 
+#include <stdio.h>
+#include <unistd.h>
+
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
 namespace hyped::io {
 
-Can::Can()
+Can::Can(core::ILogger &log_) : log_(log_), processors_()
+{
+}
+
+io::CanResult Can::initialise(const std::string &can_network_interface)
+{
+  socket_ = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+  if (socket_ < 0) {
+    log_.log(core::LogLevel::kFatal, "Unable to open CAN socket");
+    return io::CanResult::kFailure;
+  }
+  sockaddr_can socket_address;
+  socket_address.can_family  = AF_CAN;
+  socket_address.can_ifindex = if_nametoindex(can_network_interface.c_str());
+  if (socket_address.can_ifindex == 0) {
+    log_.log(core::LogLevel::kFatal, "Unable to find CAN1 network interface");
+    close(socket_);
+    socket_ = -1;
+    return io::CanResult::kFailure;
+  }
+  const int bind_status
+    = bind(socket_, reinterpret_cast<sockaddr *>(&socket_address), sizeof(socket_address));
```

Comment:
maybe put a comment on why we do the `reinterpret_cast` here? I believe it's an acceptable way to do it but it's non-obvious. I had to look up the `bind` man page.  https://www.man7.org/linux/man-pages/man2/bind.2.html